src/cluster/managers/moderation/TimeoutManager.ts

```diff
@@ -0,0 +1,90 @@
+import { TimeoutResult, UnTimeoutResult } from '@blargbot/cluster/types';
+import { humanize } from '@blargbot/cluster/utils';
+import { Guild, Member, User } from 'eris';
+import moment, { Duration } from 'moment-timezone';
+
+import { ModerationManager } from '../ModerationManager';
+import { ModerationManagerBase } from './ModerationManagerBase';
+
+export class TimeoutManager extends ModerationManagerBase {
+    private readonly ignoreTimeouts: Set<`${string}:${string}`>;
+    private readonly ignoreUnTimeouts: Set<`${string}:${string}`>;
+
+    public constructor(manager: ModerationManager) {
+        super(manager);
+        this.ignoreTimeouts = new Set();
+        this.ignoreUnTimeouts = new Set();
+    }
+
+    public async timeout(member: Member, moderator: User, authorizer: User, duration: Duration, reason?: string): Promise<TimeoutResult> {
+        const guild = member.guild;
+        const result = await this.tryTimeoutUser(guild, member.id, moderator, authorizer, duration, reason);
+        if (result !== 'success') {
+            if (typeof result === 'string')
+                return result;
+            throw result.error;
+        }
+
+        await this.modLog.logTimeout(guild, member.user, duration, moderator, reason);
+
+        return 'success';
+    }
+
+    public async removeTimeout(member: Member, moderator: User, authorizer: User, reason?: string): Promise<UnTimeoutResult> {
+        if (member.communicationDisabledUntil === null)
+            return 'notTimedOut';
+
+        const guild = member.guild;
+
+        const self = guild.members.get(this.cluster.discord.user.id);
+        if (self?.permissions.has('moderateMembers') !== true)
+            return 'noPerms';
+
+        const permMessage = await this.checkModerator(guild, undefined, authorizer.id, 'moderateMembers', 'timeoutoverride');
+        if (permMessage !== undefined)
+            return permMessage;
+
+        this.ignoreUnTimeouts.add(`${guild.id}:${member.id}`);
+        await guild.editMember(member.id, { communicationDisabledUntil: null }, `[${humanize.fullName(moderator)}] ${reason ?? ''}`);
+        await this.modLog.logUnTimeout(guild, member.user, moderator, reason);
+
+        return 'success';
+    }
+
+    private async tryTimeoutUser(guild: Guild, userId: string, moderator: User, authorizer: User, duration: Duration, reason?: string): Promise<TimeoutResult | { error: unknown; }> {
```

Comment:
Not an issue if we leave it like this, but the try pattern isnt needed for timeouts. Its there for bans because massban is a thing. Up to you if you leave it here

Reply:
I actually find it pretty useful to separate the actual action into a private method. We don't need to overcomplicate the public `timeout` method. I think I will let the method but rename it.

Should I keep the ignoreTimeouts and ignoreTimeoutClears though?